### PR TITLE
#[152454553 ][This PR combines features of Challenge two.]

### PR DESCRIPTION
**Pull request by EuniceGatehi**
**What does this PR do?**
- Creates classes, objects and methods recipes and their categories to achieve CRUD operations
- Includes comments and docstrings to code for good coding practice
- Implements the technical feedback received from my LFA's(both for week 1 and 2)
- Obtains badges for build passing[Build Status](https://travis-ci.org/EuniceGatehi/Yummy_Recipes) and coveralls [Coverage Status](https://coveralls.io/github/EuniceGatehi/Yummy_Recipes?branch=developchallengetwo) through Travis integration
- Deploys app to Heroku

**Description of Task to be completed?**
As indicated in the challenge 2 expectations:
- Setup unit testing libraries and ensure minimal tests written, to practice Test-Driven Development.
- Integrate TravisCI for Continuous Integration in your repository (with ReadMe badge).
- Integrate test coverage reporting (e.g. Coveralls) with a badge in the ReadMe.
- Obtain CI badges (e.g. from Code Climate and Coveralls) and add to ReadMe.
- Integrate the UI templates developed in challenge 1 into your Flask application and make them accessible when you run your app. 
- Develop all required features including their user interfaces. Use OOP and no database (data should be lost when the server is restarted).
- Create a pull request to your development branch and request two of your friends to review it.
- Deploy your Flask application to Heroku.

**How should this be manually tested?**
For the CRUD operations, run https://yummyflask.herokuapp.com/ to see if all features work correctly
For the readme badges, on my developchallengetwo branch, navigate to the README.md file and view the badges
**What are the relevant pivotal tracker stories?**
[Finishes #152454553](https://www.pivotaltracker.com/story/show/152454553)
[Finishes #152062326](https://www.pivotaltracker.com/story/show/152062326)
[Finishes #152181931](https://www.pivotaltracker.com/story/show/152181931)
[Finishes #152181934](https://www.pivotaltracker.com/story/show/152181934)

**Screenshots (if appropriate)**
Images for the Travis and Coveralls reports
image
image
Screenshot for my Heroku app as it appears when you click this link
https://yummyflask.herokuapp.com/
image

### CODE REVIEW

**Comment by Sylvance** [comment]
Looks alright. Coverage is good too.

**Review by jonathankamau — CHANGES_REQUESTED** [review]
Make sure you have a new line at the end of your files

File: README.md [hunk 1]
```diff
@@ -1,2 +1,48 @@
...
+
+$  (envname) pip install nose
+
+# Running the program
+
+Run the program by typing the command in your terminal : 
+
+$  (envname) python run.py
+
+and voila!you can now use the application.
```
Comment (jonathankamau) [inline README.md]: I do not see the procedure to run tests for your application

File: app/__init__.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+""" app/__init__.py """
+
+from flask import Flask
+from app.userclass import User
+from app.recipecategories import RecipecategoryClass
+from app.recipes import RecipesClass
+
+
```
Comment (jonathankamau) [inline app/__init__.py]: why this space?

File: app/recipecategories.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
+""" recipecategories.py """
+import re
+
+
```
Comment (jonathankamau) [inline app/recipecategories.py:4]: why this space?

File: app/recipes.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
+"""recipes.py"""
+import re
+
+
```
Comment (jonathankamau) [inline app/recipes.py:4]: why this space?

File: app/recipes.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                            'name': recipe_name,
+                        }
+                        item.update(edit_dict)
+                else:
+                    return "Recipe name already exists"
+        return self.owner_recipes(user, category_name)
+
+    def delete_recipe(self, recipe_name, user, category_name):
+        """Handles deletion of recipe category recipes
+
```
Comment (jonathankamau) [inline app/recipes.py:72]: why this space?

**Review by jonathankamau — COMMENTED** [review]
Why are your bootstrap css and html files on this PR? Your PR for challenge 1 should have contained these files so that once you merge them, the following PR's you create will not have these files. I expected to see only files for challenge 2 on this PR.

File: app/recipecategories.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        """
+        # Check for special characters
+        if re.match("^[a-zA-Z0-9 _]*$", category_name):
+            # Get users recipe categories
+            my_recipe_categories = self.get_owner(user)
+            # check if category name exists
+            for item in my_recipe_categories:
+                if category_name == item['name']:
+                    return "recipe category name already exists."
+            recipe_dict = {
```
Comment (jonathankamau) [inline app/recipecategories.py]: `recipe_dict`??? why not have it as `recipe_details`

File: app/recipes.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                error message or a list of recipes
+        """
+        # Get users recipes
+        my_recipes = self.owner_recipes(user, category_name)
+        for item in my_recipes:
+            if item['category'] == category_name:
+                if item['name'] != recipe_name:
+                    if item['name'] == org_recipe_name:
+                        del item['name']
+                        edit_dict = {
```
Comment (jonathankamau) [inline app/recipes.py]: `edit_dict` :flushed:

**Pull request merged**